### PR TITLE
pkg/rpc: defeat MITM attacks in most cases

### PR DESCRIPTION
**Pull request by jtolds**
This change adds a trusted registry (via the source code) of node address to node id mappings (currently only for well known Satellites) to defeat MITM attacks to Satellites. It also extends the uplink UI such that when entering a satellite address by hand, a node id prefix can also be added to defeat MITM attacks with unknown satellites.

When running `uplink setup`, satellite addresses can now be of the form `12EayRS2V1k@us-central-1.tardigrade.io` (not even using a full node id) to ensure that the peer contacted is the peer that was expected. When using a known satellite address, the known node ids are used if no override is provided.

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: pkg/rpc/dial.go [hunk 1]
```diff
@@ -111,6 +114,54 @@ func (d Dialer) DialAddressID(ctx context.Context, address string, id storj.Node
 	return d.dial(ctx, address, d.TLSOptions.ClientTLSConfig(id))
 }
 
+// DialAddressInsecureBestEffort is like DialAddressInsecure but tries to dial a node securely if
+// it can.
+//
+// nodeURL is like a storj.NodeURL but (a) requires an address and (b) does not require a
```
Comment (egonelbre) [inline pkg/rpc/dial.go]: `storj.NodeURL` doesn't have any restriction on the address nor URL. You could still use `storj.NodeURL` here and then check whether `if nodeurl.Address == "" { return Error.New }`

File: pkg/peertls/tlsopts/tls.go [hunk 2]
```diff
@@ -48,6 +49,12 @@ func (opts *Options) ClientTLSConfig(id storj.NodeID) *tls.Config {
 	return opts.tlsConfig(false, verifyIdentity(id))
 }
 
+// ClientTLSConfigPrefix returns a TSLConfig for use as a client in handshaking with a peer.
+// The peer node id is validated to match the given prefix
+func (opts *Options) ClientTLSConfigPrefix(idPrefix string) *tls.Config {
+	return opts.tlsConfig(false, verifyIdentityPrefix(idPrefix))
```
Comment (egonelbre) [inline pkg/peertls/tlsopts/tls.go]: Why verify the prefix instead of the whole ID?

File: pkg/rpc/dial.go (hunk 1 above)
Comment (jtolds) [reply pkg/rpc/dial.go]: i believe it uses a storj.NodeID, which i would hope has a length check. this has no length check, so i didn't want to use a storj.NodeID for a node id prefix

File: pkg/peertls/tlsopts/tls.go (hunk 2 above)
Comment (jtolds) [reply pkg/peertls/tlsopts/tls.go]: better UX. you can have someone type just 12vha9oTFnerx@33.20.0.1:7777 instead of something huge

**Comment by stefanbenten** [comment]
We should add those changes to our docs!

File: pkg/rpc/dial.go (hunk 1 above)
Comment (egonelbre) [reply pkg/rpc/dial.go]: ah, missed the prefix part

**Comment by jtolds** [comment]
Will add lots! of documentation with our push to add community satellites!

**Pull request merged**